File: MFBobTests/iOS/AutomationAppIos/app/AutomationAppIos/AutomationAppIos/SendHeadersTest.swift
//
//  SendHeadersTest.swift
//  AutomationAppIos
//
//  Created by Amitai Madar on 13/03/16.
//  Copyright © 2016 Amitai Madar. All rights reserved.
//

import Foundation

#if WATCH
    import IBMMobileFirstPlatformFoundationWatchOS
#endif
#if PHONE
    import IBMMobileFirstPlatformFoundation
#endif


internal class SendHeadersTest : AutomaticTest {
    
   var type:String? = "string"; //string; hash; json; byte; error
    
    internal override func run(urlParams :[(String, String)]) {
        let dictParams :NSMutableDictionary =  NSMutableDictionary()
        for (_, value) in urlParams.enumerate() {
            dictParams.setValue(value.1 as String, forKey: value.0 as String)
        }
        type = dictParams.valueForKey("type") as? String
        var adapterPath:NSURL?
        adapterPath = NSURL(string: "http://httpbin.org/headers")
        let request: WLResourceRequest = WLResourceRequest(URL: adapterPath, method: "GET");
        
        switch (type!){
            case "setHeaders":
                request.setHeaderValue("header1 , header2", forName: "testHeader")
                break;
            case "addHeader":
                request.addHeaderValue("header1", forName: "Testheader1")
                request.addHeaderValue("header2", forName: "Testheader2")
                break;
            case "addSameHeader":
                request.addHeaderValue("header1", forName: "Testh1")
                request.addHeaderValue("header2", forName: "Testh1")
                break;
            case "removeHeaders":
                request.addHeaderValue("header1", forName: "Testheader1")
                request.addHeaderValue("header2", forName: "Testheader2")
                request.removeHeadersForName("Testheader2");
                break;
            case "getAllHeaders":
                request.addHeaderValue("header1", forName: "Testheader1")
                request.addHeaderValue("header2", forName: "Testheader2")
                var headerMap =  request.headers
                var count  = 0;
                for (var i = 0; i < headerMap.count; i++) {
                    let e =  headerMap[i]
                    let key = e.valueForKey("headerName") as! String
                    if (key == ("Testheader1") || key == ("Testheader2")){
                        count++;
                    }
                    var headerList =  request.headersForName(key)
                    for (var i = 0; i < headerList.count; i++) {
                        let headerVal =  headerList[i] as! String
                        if (headerVal == "header1" || headerVal == "header2"){
                            count++
                        }
                    }
                }
                if (count == 4) {
                    GlobalVar.result = "Success";
                }
                else {
                    GlobalVar.result = "Failure: missing header";
                }
                break;
            case "getHeaders":
                request.addHeaderValue("header1", forName: "Testheader1")
                request.addHeaderValue("header2", forName: "Testheader1")
                var headerList =  request.headersForName("Testheader1")
                var count  = 0;
                for (var i = 0; i < headerList.count; i++) {
                    let headerVal =  headerList[i] as! String
                    if (headerVal == "header1" || headerVal == "header2"){
                        count++
                    }
                }
                if (count == 2) {
                    GlobalVar.result = "Success";
                }
                else {
                    GlobalVar.result = "Failure: missing header";
                }
                break;
            default:
                GlobalVar.result = "Failure, no such type! ";
                return;
            }
            request.sendWithCompletionHandler(completionHandler)
        }
    
    
    internal func completionHandler(response:WLResponse!, error:NSError!) {
        let responseText = response.responseText
        if((error) != nil){
            GlobalVar.result = "Failure" + GlobalVar.gettErrorMessage(error)
        }
        else{
            if(type == "setHeaders" && responseText.containsString("Testheader") &&
                    responseText.containsString("header1") &&
                    responseText.containsString("header2")) {
                        GlobalVar.result = "Success";
            }
            else if(type == "addHeader" && responseText.containsString("Testheader1") &&
                    responseText.containsString("Testheader2") &&
                    responseText.containsString("header1") &&
                    responseText.containsString("header2")) {
                        GlobalVar.result = "Success";
            }
            else if(type == "addSameHeader" && responseText.containsString("Testh1") &&
                    responseText.containsString("header1") &&
                    responseText.containsString("header2")) {
                        GlobalVar.result = "Success";
            }
            else if (type == "removeHeaders" && responseText.containsString("Testheader1") &&
                    responseText.containsString("header1")) {
                        GlobalVar.result = "Success";
            }
            else if (type == "getHeaders" || type == "getAllHeaders") {
                //ok
            }
            else{
                GlobalVar.result = "Failure: send is different from received"
            }
            
        }
        WLClient.sharedInstance().setServerUrl( WLClient.sharedInstance().serverUrl())
    }
}

File: MFBobTests/Android/com.example.amitaim.automationapp/app/AutomationApp/app/src/main/java/com/example/amitaim/automationapp/Tests/SendHeadersTest.java
package com.example.amitaim.automationapp.Tests;


import com.example.amitaim.automationapp.AndroidChallengeHandler;
import com.example.amitaim.automationapp.MainActivity;
import com.worklight.wlclient.api.WLFailResponse;
import com.worklight.wlclient.api.WLResourceRequest;
import com.worklight.wlclient.api.WLResponse;
import com.worklight.wlclient.api.WLResponseListener;
import com.worklight.wlclient.auth.WLAuthorizationManagerInternal;

import org.json.JSONObject;

import java.net.URI;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import fi.iki.elonen.NanoHTTPD;

/**
 * Created by amitaim on 03/01/16.
 */
public class SendHeadersTest extends AutomaticTest {

    private String type; //string; hash; json; byte; error
    JSONObject testJson = new JSONObject();
    byte[] testBytes;

    public SendHeadersTest() {
        type = "string";
    }

    public SendHeadersTest(NanoHTTPD.IHTTPSession session) {
        type = session.getParms().get("type");
    }

    @Override
    public void run() {
        try {
            URI adapterPath ;
            adapterPath = new URI("http://httpbin.org/headers");
            WLResourceRequest request = new WLResourceRequest(adapterPath, WLResourceRequest.GET);

            switch (type){
                case "setHeaders":
                    List<String> headersVals = new ArrayList<String>();
                    headersVals.add("header1");
                    headersVals.add("header2");
                    HashMap<String, List<String>> headers = new HashMap<String, List<String>>();
                    headers.put("testheader", headersVals);
                    request.setHeaders(headers);
                    break;
                case "addHeader":
                    request.addHeader("Testheader1", "header1");
                    request.addHeader("Testheader2", "header2");
                    break;
                case "addSameHeader":
                    request.addHeader("Testheader1", "header1");
                    request.addHeader("Testheader1", "header2");
                    break;
                case "removeHeaders":
                    request.addHeader("Testheader1", "header1");
                    request.addHeader("Testheader2", "header2");
                    request.removeHeaders("Testheader2");
                    break;
                case "getAllHeaders":
                    request.addHeader("Testheader1", "header1");
                    request.addHeader("Testheader2", "header2");
                    Map<String, List<String>> headerMap =  request.getAllHeaders();
                    int count  = 0;
                    for (Map.Entry<String, List<String>> e : headerMap.entrySet()) {
                        if (e.getKey().equals("Testheader1")) count ++;
                        if (e.getKey().equals("Testheader2")) count ++;
                        for (String headerValue : e.getValue()) {
                            if (headerValue.equals("header1")) count ++;
                            if (headerValue.equals("header2")) count ++;
                        }
                    }
                    if (count == 4) {
                        MainActivity.AutomationServer.result = "Success";
                    }
                    else {
                        MainActivity.AutomationServer.result = "Failure: missing header";
                    }
                    break;
                case "getHeaders":
                    request.addHeader("Testheader1", "header1");
                    request.addHeader("Testheader1", "header2");
                    List<String> headerList =  request.getHeaders("Testheader1");
                    count  = 0;
                    for (int i = 0; i < headerList.size(); i++) {
                        String headerVal =  headerList.get(i);
                        if (headerVal.equals("header1")) count ++;
                        if (headerVal.equals("header2")) count ++;

                    }
                    if (count == 2) {
                        MainActivity.AutomationServer.result = "Success";
                    }
                    else {
                        MainActivity.AutomationServer.result = "Failure: missing header";
                    }
                    break;
                default:
                    MainActivity.AutomationServer.result = "Failure, no such type! ";
                    return;
            }
            request.send(new MyResponseListener());

        }catch (Exception e){
            MainActivity.AutomationServer.result = "Failure " + e.getMessage();
            return;
        }
    }

    public class MyResponseListener implements WLResponseListener {

        @Override
        public void onSuccess(WLResponse response) {
            if(type.equals("setHeaders")) {
                if (response.getResponseText().contains("Testheader") &&
                        response.getResponseText().contains("header1") &&
                        response.getResponseText().contains("header2")) {
                    MainActivity.AutomationServer.result = "Success";
                }
            }
            else if(type.equals("addHeader")) {
                if (response.getResponseText().contains("Testheader1") &&
                        response.getResponseText().contains("Testheader2") &&
                        response.getResponseText().contains("header1") &&
                        response.getResponseText().contains("header2")) {
                    MainActivity.AutomationServer.result = "Success";
                }
            }
            else if(type.equals("addSameHeader")) {
                if (response.getResponseText().contains("Testheader1") &&
                        response.getResponseText().contains("header1") &&
                        response.getResponseText().contains("header2")) {
                    MainActivity.AutomationServer.result = "Success";
                }
            }
            else if (type.equals("removeHeaders")) {
                if (response.getResponseText().contains("Testheader1") &&
                        response.getResponseText().contains("header1")) {
                    MainActivity.AutomationServer.result = "Success";
                }
            }
            else if (type.equals("getHeaders") || type.equals("getAllHeaders")) {
                //ok
            }
            else
                MainActivity.AutomationServer.result = "Failure: send is different from received";
            WLAuthorizationManagerInternal.getInstance().clearRegistration();
            return;
        }

        @Override
        public void onFailure(WLFailResponse response) {
            String errorMsg=response.getErrorMsg();
            if (errorMsg != null)
                errorMsg=errorMsg.replace("\n","").replace("\r","").replace("\t","");
            MainActivity.AutomationServer.result = "Failure " + errorMsg;
            WLAuthorizationManagerInternal.getInstance().clearRegistration();
            return;
        }
    }
}
